### ISSUE TITLE
Error types should be part of nakama-common

### ISSUE THREAD

**Issue by lauri-skunkworks** [heroiclabs/nakama#691]
## Description
Referring to https://forum.heroiclabs.com/t/dependency-on-server-module-does-not-work/1982

The errors returned by various APIs of the server should be moved from nakama (server) module into nakama-common. The error types should be part of the server's public API that can be utilized by the plugin code.

For example, the plugin code might want to differentiate if creating a group failed because the suggested group name was already in use or because of some other error. This should be possible by checking the actual error type instead of inspecting the human readable error message (which might change in the future).

The change covers a set of error variables in various source code files e.g. ErrTournamentNotFound (core_tournament.go), ErrGroupNameInUse (core_group.go), etc.

## Steps to Reproduce
1. Try to differentiate type safely the different errors returned by Nakama's server APIs.

## Expected Result
I'm able to import a package where the error types are defined so that I can compare a runtime error to the existing error types without having to do string comparison with the the human readable error message.

    group, err := nk.createGroup(...)
    if errors.Is(err, server.ErrGroupNameInUse) {
        // Handle this type of error in a special way
    } else {
        // Handle other errors
    }

## Actual Result
I cannot differentiate between the error types in a type safe way because I cannot import the server module in my plugin (see referenced forum post).

## Your Environment
- Nakama: 3.5.0

**Comment by mnarimani** [comment]
Agreed. Also I think having a custom error for batch operations is really useful. 
For example, If 5 storage writes fail, it's good to know which object failed. 

I can work on a PR for this if Heroic Labs accepts PRs.
@novabyte

**Comment by novabyte** [comment]
@pnarimani There's a small number of error types which could be moved to the Nakama common package to be consumed by Go runtime code. It would be great to have a pull request. 

I think separate it out from the addition of any new error types which are used to detect situations like the partial write error you describe. I think that case will be trickier than you foresee because of the transactional guarantees and performance requirements of that code path.

**Comment by pnarimani** [comment]
> I think that case will be trickier than you foresee because of the transactional guarantees and performance requirements of that code path.

Is there a benchmark that I need to meet?
